[v2] Delay page transition after load

pieh opened the pull request:
closes #3372

This introduces new browser APIs:
 - `onFetchingResources` (to replace `window.___emitter.on("onDelayedLoadPageResources", () => {})`)
 - `onFetchedResources` (to replace `window.___emitter.on("onPostLoadPageResources", () => {})`)

Update nprogress plugin to use those new APIs instead of gatsby internals

KyleAMathews commented:
BTW, the issue with https://github.com/gatsbyjs/gatsby/issues/3372 is a bit different — the problem there is that we try to navigate at 1 second even though there's nothing to navigate — in retrospect, I have no idea what I was thinking there 😅 https://github.com/gatsbyjs/gatsby/issues/3372#issuecomment-384008668

On the names, how about keeping the existing emitter names? `onDelayedLoadPageResources`? This follows the API naming spec https://www.gatsbyjs.org/docs/api-specification/#extension-apis

The process is `loadPageResources` so we could have `onPreLoadPageResources`, `onDelayedLoadPageResources`, and `onPostLoadPageResources`.

pieh commented:
I removed line that was navigating after 1s ( https://github.com/gatsbyjs/gatsby/pull/5109/files#diff-2d21ea42ec874a0988977e57b17251aaL93 ). I also added so we emit failed resource fetches so we don't wait indefenitely for them (but this could probably need more work than just emiting fails only if page is not in manifest)

pieh commented:
And I have no attachment to my names - I will rename them.

KyleAMathews commented:
>I removed line that was navigating after 1s

Ah ok, hadn't read the diff :-)

KyleAMathews commented on packages/gatsby/src/utils/api-browser-docs.js:
```diff
@@ -8,6 +8,24 @@
 exports.onClientEntry = true
 
+/**
+ * Called when gatsby is fetching resources for a page longer than 1 second.
```
Capitalize Gatsby

KyleAMathews commented on packages/gatsby/src/utils/api-browser-docs.js:
```diff
@@ -8,6 +8,24 @@
+exports.onFetchingResources = true
+
+/**
+ * Called when gatsby has fetched all resources for page
```
Same here

pieh commented:
Just double checking new browser APIs:
- `onPreLoadPageResources` - triggered when `navigateTo` is called - I didn't have this hook, to not encourage showing loading indicators prematurely, but sure can add this
- `onDelayedLoadPageResources` - triggered 1s after `navigateTo` is called - with this PR we don't actually do anything after 1s other than triggering this hook, so if we would add  `onPreLoadPageResources` then maybe we can let plugin handle delayed indicator showing and not actually add this API? 
- `onLoadPageResources` - triggered after all resources are loaded

KyleAMathews commented:
>I didn't have this hook, to not encourage showing loading indicators prematurely, but sure can add this

I want to encourage people to use the default 1 second as there's a lot of research saying that's the right time to show a progress indicator — but there's perhaps legitimate reasons for a lifecycle hook at the start so add it as well.

pieh commented:
updated with renamed/added APIs:
- `onPreLoadPageResources` - triggered when `navigateTo` is called
- `onDelayedLoadPageResources` - triggered 1s after `navigateTo` is called (if didn't finish yet)
- `onLoadPageResources` - triggered after all resources are loaded

KyleAMathews commented:
So... was reading through this and I don't think our API names are right still. These APIs are all around loading resources _as part of_ page navigation. We might for example want in the future to expose lifecycle hooks explicitly around loading resources which of course happens outside of route changes.

So we should tie the API names around route changes.

Perhaps:

* `onPreRouteUpdate`
* `onLoadPageResourcesDelayed`

We already have [`onRouteUpdate`](https://www.gatsbyjs.org/docs/browser-apis/#onRouteUpdate) so perhaps we don't need `onLoadPageResources` (until we add life cycle events around actual page resource loading).

Thoughts?

pieh commented:
Pivoting to route changes defenitely makes sense. Usecase for those new APIs was around displaying loading indicators and that really only makes sense when we load resources because of navigation and not prefetching. 

Codewise not much would change apart from renaming APIs (and I think adding `onPreRouteUpdate` to `develop` mode for develop/build consistency).

I will give it more thought over the weekend, but just checking if we are on the same page:
> Perhaps:
> 
> onPreRouteUpdate
> onLoadPageResourcesDelayed

Shouldn't second one be `onDelayedRouteUpdate` or `onRouteUpdateDelayed`?

m-allanson commented:
Let's go with `onRouteUpdateDelayed`

pieh commented:
updated PR with browser APIs:
 - `onRouteUpdateDelayed` - this is only called in production
 - `onPreRouteUpdate` - this is in both dev and production (both when using `navigateTo` and when using history directly - for example browser's "go back")
 - `onRouteUpdateError` - this is new one (not discussed earlier). It's called when we try to navigate to path that gatsby doesn't know about (currently only in production) - I think we need this if we want to use `onRouteUpdate` "family" APIs to handle loading indicators. I was thinking about passing `error` to `onRouteUpdate` but decided against it as this is behaviour change that might go unnoticed (even if noted in changelog) and cause problems when upgrading to v2.

m-allanson commented on packages/gatsby/src/utils/api-browser-docs.js:
```diff
@@ -17,6 +17,45 @@ exports.onClientEntry = true
+ * @param {object} $0.action The "action" that caused the route change
+ * @param {object} $0.error An error object
+ * @example
+ * exports.onRouteUpdateError = ({ location, error }) => {
```
Should the examples include every param? e.g. adding `action` to this one `({ location, action, error })`

pieh replied:
I checked other examples and `onRouteUpdate` (that I didn't change at all) doesn't use all params ( https://github.com/pieh/gatsby/blob/5007dbffcfe9d2985f56997412362e20be728c66/packages/gatsby/src/utils/api-browser-docs.js#L65 ). But given that we only console log in those examples it's no problem to add all params there.

m-allanson replied:
oh ok, maybe leave this as-is if that's how it's done elsewhere.

pieh commented:
> onRouteUpdateError - this is new one (not discussed earlier). It's called when we try to navigate to path that gatsby doesn't know about (currently only in production) - I think we need this if we want to use onRouteUpdate "family" APIs to handle loading indicators. I was thinking about passing error to onRouteUpdate but decided against it as this is behaviour change that might go unnoticed (even if noted in changelog) and cause problems when upgrading to v2.

TO-DO: remove that one, there is no clear use case for this.

pieh commented:
I was working on this to get it in mergable shape but after more intensive testing I found some issues that I would like to resolve first:
 - https://github.com/gatsbyjs/gatsby/blob/v2/packages/gatsby/src/cache-dir/loader.js#L338-L388 this whole part is weird and seems like a result of one of previous master->v2 merges - ie we shouldn't call `getResourceModule` if resources are already in `pathScriptsCache`
 - navigating to not existing path would cause multiple requests for 404 page data:
![404requests](https://user-images.githubusercontent.com/419821/41381948-3100c3b0-6f6a-11e8-80d6-8a00b405bd08.png)
